File: langs/python/src/dotlyte/cli/diff.py
"""``dotlyte diff`` — Compare two .env files."""

from __future__ import annotations

import re
import sys
from pathlib import Path

from dotlyte.masking import REDACTED, _SENSITIVE_PATTERNS


def run_diff(file1: str, file2: str) -> None:
    """Show added, removed, and changed keys between two .env files.

    Args:
        file1: Path to the first .env file.
        file2: Path to the second .env file.

    """
    path1 = Path(file1).resolve()
    path2 = Path(file2).resolve()

    if not path1.is_file():
        print(f"❌ File not found: {file1}", file=sys.stderr)
        sys.exit(1)
    if not path2.is_file():
        print(f"❌ File not found: {file2}", file=sys.stderr)
        sys.exit(1)

    env1 = _parse_env_file(path1.read_text(encoding="utf-8"))
    env2 = _parse_env_file(path2.read_text(encoding="utf-8"))

    all_keys = sorted(set(env1) | set(env2))

    added: list[str] = []
    removed: list[str] = []
    changed: list[tuple[str, str, str]] = []
    unchanged: list[str] = []

    for key in all_keys:
        in1 = key in env1
        in2 = key in env2
        if not in1 and in2:
            added.append(key)
        elif in1 and not in2:
            removed.append(key)
        elif env1[key] != env2[key]:
            changed.append((key, _mask(key, env1[key]), _mask(key, env2[key])))
        else:
            unchanged.append(key)

    print(f"\n📊 dotlyte diff\n")
    print(f"  {file1} ↔ {file2}\n")

    if added:
        print(f"  ➕ Added ({len(added)}):")
        for k in added:
            print(f"     + {k}={_mask(k, env2[k])}")

    if removed:
        print(f"  ➖ Removed ({len(removed)}):")
        for k in removed:
            print(f"     - {k}={_mask(k, env1[k])}")

    if changed:
        print(f"  ✏️  Changed ({len(changed)}):")
        for k, old, new in changed:
            print(f"     ~ {k}: {old} → {new}")

    print(
        f"\n  Summary: {len(added)} added, {len(removed)} removed, "
        f"{len(changed)} changed, {len(unchanged)} unchanged\n"
    )


# ──── Internal helpers ────


def _parse_env_file(content: str) -> dict[str, str]:
    """Parse a .env file into a key-value dict."""
    result: dict[str, str] = {}
    for line in content.splitlines():
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            continue
        stripped = trimmed[7:].strip() if trimmed.startswith("export ") else trimmed
        eq_idx = stripped.find("=")
        if eq_idx > 0:
            key = stripped[:eq_idx].strip()
            value = stripped[eq_idx + 1 :].strip()
            if (
                len(value) >= 2
                and value[0] == value[-1]
                and value[0] in ('"', "'")
            ):
                value = value[1:-1]
            result[key] = value
    return result


def _is_sensitive(key: str) -> bool:
    """Return True if *key* matches a built-in sensitive pattern."""
    for pattern in _SENSITIVE_PATTERNS:
        if pattern.search(key):
            return True
    return False


def _mask(key: str, value: str) -> str:
    """Mask value if the key looks sensitive or the value looks like a secret."""
    if _is_sensitive(key):
        return REDACTED
    if len(value) > 40 and re.match(r"^[A-Za-z0-9+/=_-]+$", value):
        return value[:8] + "..." + value[-4:]
    return value

File: langs/javascript/src/cli/commands/diff.ts
/**
 * `dotlyte diff` — Compare two env files.
 *
 * Usage:
 *   dotlyte diff <file1> <file2>
 *
 * Shows added, removed, and changed keys. Sensitive values are masked.
 */

import { readFileSync, existsSync } from "node:fs";
import { resolve } from "node:path";
import { isAutoSensitive, REDACTED } from "../../masking.js";

export async function diff(args: string[]): Promise<void> {
  if (args.length < 2) {
    throw new Error("Usage: dotlyte diff <file1> <file2>");
  }

  const file1 = resolve(args[0]!);
  const file2 = resolve(args[1]!);

  if (!existsSync(file1)) throw new Error(`File not found: ${file1}`);
  if (!existsSync(file2)) throw new Error(`File not found: ${file2}`);

  const env1 = parseEnvFile(readFileSync(file1, "utf-8"));
  const env2 = parseEnvFile(readFileSync(file2, "utf-8"));

  const allKeys = new Set([...Object.keys(env1), ...Object.keys(env2)]);

  const added: string[] = [];
  const removed: string[] = [];
  const changed: { key: string; from: string; to: string }[] = [];
  const unchanged: string[] = [];

  for (const key of allKeys) {
    const in1 = key in env1;
    const in2 = key in env2;

    if (!in1 && in2) {
      added.push(key);
    } else if (in1 && !in2) {
      removed.push(key);
    } else if (env1[key] !== env2[key]) {
      changed.push({
        key,
        from: maskIfSensitive(key, env1[key]!),
        to: maskIfSensitive(key, env2[key]!),
      });
    } else {
      unchanged.push(key);
    }
  }

  console.log(`\n📊 dotlyte diff\n`);
  console.log(`  ${args[0]} ↔ ${args[1]}\n`);

  if (added.length > 0) {
    console.log(`  ➕ Added (${added.length}):`);
    for (const k of added) {
      console.log(`     + ${k}=${maskIfSensitive(k, env2[k]!)}`);
    }
  }

  if (removed.length > 0) {
    console.log(`  ➖ Removed (${removed.length}):`);
    for (const k of removed) {
      console.log(`     - ${k}=${maskIfSensitive(k, env1[k]!)}`);
    }
  }

  if (changed.length > 0) {
    console.log(`  ✏️  Changed (${changed.length}):`);
    for (const c of changed) {
      console.log(`     ~ ${c.key}: ${c.from} → ${c.to}`);
    }
  }

  console.log(`\n  Summary: ${added.length} added, ${removed.length} removed, ${changed.length} changed, ${unchanged.length} unchanged\n`);
}

function parseEnvFile(content: string): Record<string, string> {
  const result: Record<string, string> = {};
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    const stripped = trimmed.startsWith("export ") ? trimmed.slice(7).trim() : trimmed;
    const eqIdx = stripped.indexOf("=");
    if (eqIdx > 0) {
      const key = stripped.slice(0, eqIdx).trim();
      let value = stripped.slice(eqIdx + 1).trim();
      // Remove quotes
      if (value.length >= 2 && value[0] === value[value.length - 1] && (value[0] === '"' || value[0] === "'")) {
        value = value.slice(1, -1);
      }
      result[key] = value;
    }
  }
  return result;
}

function maskIfSensitive(key: string, value: string): string {
  if (isAutoSensitive(key)) {
    return REDACTED;
  }
  // Also mask long values that look like secrets
  if (value.length > 40 && /^[A-Za-z0-9+/=_-]+$/.test(value)) {
    return value.slice(0, 8) + "..." + value.slice(-4);
  }
  return value;
}
